### PR TITLE
Split coremodels into coremodel and kubecontroller

### LINKED ISSUE #46674: Make exporting a `LineageFactory` func part of the coremodel component's responsibility

### PR DESCRIPTION
**Pull request by sdboyer**
**What this PR does / why we need it**:

This separates out k8s-related bits from coremodels. Everything k8s-related is now expected to be in a `crd` subdir of a coremodel component, and wrapped up into a new kind of thing  called a `KubeModel`. Kubemodels are then layered on top of the coremodel itself, meaning the coremodel is really just the pure Thema declaration. (i've decided it's not worth the extra complexity of supporting a theoretical Go-as-canonical-schema path when there are such clear shortcomings.)

There's a fair bit of renaming here, and there's more cleanup that could be done over in e.g. `pkg/schema`, but i think we can tidy more as we iterate.

**Which issue(s) this PR fixes**:

Fixes #46674

**Special notes for your reviewer**:

Couple small TODOs

* [x] Update compgen.yaml rules accordingly
* [x] Mark purely intent api/k8s-related parts as owned by @'grafana/cloud-middleware 
* [x] Fully split k8s concerns into subpackages, including frameworky bits

### CODE REVIEW

**Comment by sdboyer** [comment]
UUUUUUUUUUUGHHHHH i hadn't updated my local intent-api branch

will fix the conflicts tomorrow

**Comment by sdboyer** [comment]
OK, it's clean now.

File: go.mod
Comment (radiohead) [inline go.mod:149]: Could you merge this with the other block on line 150? Strange that it didn't happen automatically - I think running `go mod tidy` might do the trick here.

File: internal/components/components.go
Comment (radiohead) [inline internal/components/components.go:14]: If we aim for clear separation of concerns, I'd move all kube-related interfaces / loaders to a subpackage, e.g. `internal/components/kube` or similar.

File: internal/components/datasource/coremodel.go
Comment (radiohead) [inline internal/components/datasource/coremodel.go:23]: I'm not a fan of this approach. Since `Lineage#Schema` returns an error, can we inject `thema.Schema` to `Coremodel` during loading, based on the version, instead of reading it here, because there's no way of dealing with errors?

Alternatively, we could change the signature of the method to return an error.

File: internal/components/datasource/schema.go
Comment (radiohead) [inline internal/components/datasource/schema.go:23]: Can this be extracted into a loader instead?

File: internal/components/kube_registry.go
Comment (radiohead) [inline internal/components/kube_registry.go:1]: Similar to `KubeModel`, it would be nice to put it into a kube-specific package.

File: internal/components/registry.go
Comment (radiohead) [inline internal/components/registry.go:55]: I don't think this belongs to the Registry. Maybe we could do this in a loader instead?

**Review by radiohead — COMMENTED** [review]
Looks good! I would provide a bit more separation of concerns by moving kube-specific machinery to a separate package and I'd also add a Thema loader and move validation / schema derivation there.

File: internal/components/kube_registry.go
Comment (sdboyer) [reply internal/components/kube_registry.go:1]: agreed, just didn't quite get there

File: internal/components/datasource/schema.go
Comment (sdboyer) [reply internal/components/datasource/schema.go:23]: Isn't it a loader already?

Comment (sdboyer) [reply internal/components/datasource/schema.go:23]: Oh wait, i see what you're saying - no, i think this is probably preferable as-is, because it's following Thema's `LineageFactory` [pattern](https://github.com/grafana/thema/blob/main/docs/go-mapping.md#goal-the-lineage-factory) - the canonical way of exposing a Thema lineage in Go.

Exposing this will eventually be important for a [public] audience that wants to directly rely on the lineage.

File: internal/components/components.go
Comment (sdboyer) [reply internal/components/components.go:14]: yeah, agreed

File: go.mod
Comment (sdboyer) [reply go.mod:149]: 👍 will do

File: internal/components/datasource/coremodel.go
Comment (sdboyer) [reply internal/components/datasource/coremodel.go:23]: It's plausible that this whole "Thema + current Go version" gets moved generically into Thema, anyway - it's already sorta there with `InputKernel`

File: internal/components/datasource/schema.go
Comment (sdboyer) [reply internal/components/datasource/schema.go:23]: Example: when i write a Go library for generating dashboards, it'll never be used by the Go backend (and therefore needs to be ergonomic/independent of wire). Instead, i'll want to call this `Lineage()` func directly.

File: internal/components/registry.go
Comment (sdboyer) [reply internal/components/registry.go:55]: i went back and forth on this a bit, but it ultimately came down to exposing the standard Thema `LineageFactory` func. As long as that's a goal, then there can't really be a single loader which both loads a lineage _and_ checks Go struct assignability, as your `loadThemaLineage()` does/did, because the logical layer of `LineageFactory` does not assume the existence of a corresponding Go struct.

If/when we genericize the clustered relationship of `Lineage`/`LineageFactory` + current version + Go struct into Thema under some new abstraction (it's sorta like a proto-[kernel](https://github.com/grafana/thema/tree/main/kernel)?) then we can put an assignability check into there, probably even generically in Thema.

Until then, though, it seemed OK to put here. Having it here provides the guarantee, "all  thema schemas in the registry are assignable to their corresponding Go structs," which is ultimately what we want, and also reasonable statement of concern separation to make. It might be better to have earlier in the process, but it's hard to imagine a case in which something escapes this check.

File: .github/CODEOWNERS
Comment (radiohead) [inline .github/CODEOWNERS:172]: Looks like this team doesn't have write access to this repo.

File: compgen.yaml
Comment (radiohead) [inline compgen.yaml:46]:
```suggestion
        name: ProvideKubeController
```

File: compgen.yaml
Comment (radiohead) [inline compgen.yaml:44]:
```suggestion
      ProvideKubeController:
```

**Review by radiohead — APPROVED** [review]
A few minor issues - we can fix them now or later, I'll let you decide.

I'd also like to re-structure Kubernetes bits a bit more, but I'll make a separate PR after you merge this one.

File: internal/components/datasource/coremodel.go
Comment (radiohead) [reply internal/components/datasource/coremodel.go]: I very much dislike panics in general, especially ones that happen at runtime due to misconfiguration. I think it's OK to leave it as it is right now, but I'd rather have it panic at build time or gracefully handle the error (e.g. skip loading malformed schema, or at least exit gracefully).

File: .github/CODEOWNERS
Comment (sdboyer) [reply .github/CODEOWNERS:172]: mmm, we'll need to get that addressed - @torkelo, if we're going to have the cloud folks run with the k8s/intent api, can we grant access? Or perhaps we should make a new, smaller team than all of @'grafana/cloud-middleware?

File: internal/components/datasource/schema.go
Comment (radiohead) [reply internal/components/datasource/schema.go]: Okay, that's fair. I would like to avoid as much boilerplate as possible, but I think this particular case should be fixed by codegen instead.

**Comment by sdboyer** [comment]
> I'd also like to re-structure Kubernetes bits a bit more, but I'll make a separate PR after you merge this one.

For sure 👍 

one thing i was thinking about was actually maybe just having all the kubecontroller components in a separate tree entirely from coremodels

File: internal/components/registry.go
Comment (radiohead) [reply internal/components/registry.go]: Sounds good. I think this is definitely something that we could / should verify either by using a Kernel or with codegen.

**Comment by radiohead** [comment]
> > I'd also like to re-structure Kubernetes bits a bit more, but I'll make a separate PR after you merge this one.
> 
> For sure 👍
> 
> one thing i was thinking about was actually maybe just having all the kubecontroller components in a separate tree entirely from coremodels

Yeah, something like that. And I'd like to also move / split Kubernetes config loading / clientset / bridge part as well - I'll look into this next week.

**Pull request merged**
